src/main/java/at/hannibal2/skyhanni/features/chat/RareDropMessages.kt

```diff
@@ -2,82 +2,155 @@ package at.hannibal2.skyhanni.features.chat
     )
 
     /**
      * REGEX-TEST: §e[NPC] Oringo§f: §b✆ §f§r§8• §fBlue Whale Pet
      * REGEX-TEST: §e[NPC] Oringo§f: §b✆ §f§r§8• §5Giraffe Pet
      */
-    private val oringoPattern by chatGroup.pattern(
-        "pet.oringopattern",
+    private val oringoPattern by petGroup.pattern(
+        "oringomessage",
         "(?<start>§e\\[NPC] Oringo§f: §b✆ §f§r§8• )§(?<rarityColor>.)(?<petName>[^§(.]+)(?<end> Pet)"
     )
 
-    private val patterns = listOf(
-        petDroppedPattern, petFishedPattern, petClaimedPattern, petObtainedPattern, oringoPattern
+    /**
+     * REGEX-TEST: §6§lRARE DROP! §r§fEnchanted Book §r§b(+§r§b208% §r§b✯ Magic Find§r§b)
+     */
+    private val enchantedBookPattern by repoGroup.pattern(
+        "enchantedbook",
+        "(?<start>(?:§.)+RARE DROP!) (?<color>(?:§.)*)Enchanted Book (?<end>§r§b\\([+](?:§.)*(?<mf>\\d*)% §r§b✯ Magic Find§r§b\\)).*"
     )
 
-    private val config get() = SkyHanniMod.feature.chat.petRarityDropMessage
+    private val petPatterns = listOf(
+        petDroppedPattern, petFishedPattern, petClaimedPattern, petObtainedPattern, oringoPattern,
+    )
+
+    private val ignoredBookIslands = setOf(
+        IslandType.DARK_AUCTION,
+        IslandType.DUNGEON_HUB,
+        IslandType.CATACOMBS,
+    )
+
+    private val userLuck get() = ProfileStorageData.playerSpecific?.limbo?.userLuck
+
+    private val config get() = SkyHanniMod.feature.chat.rareDropMessages
 
     @HandleEvent(onlyOnSkyblock = true)
     fun onChat(event: SkyHanniChatEvent) {
-        if (!config) return
+        if (!config.petRarity) return
 
-        patterns.matchMatchers(event.message) {
+        petPatterns.matchMatchers(event.message) {
             var start = group("start")
             val rarityColor = group("rarityColor")
             val rarityName = colorCodeToRarity(rarityColor.first()).uppercase()
             val petName = group("petName")
             val end = group("end")
-            if (start.endsWith("a ") && rarityName.matches("(?i)[aeiou].*".toRegex()))
-                start = start.replace(" $".toRegex(), "n ")
+            if (start.endsWith("a ") && rarityName.first().isVowel())
+                start = start.substring(0..start.length - 2) + "n "
 
             event.chatComponent = ChatComponentText(
                 "$start§$rarityColor§l$rarityName §$rarityColor$petName$end"
             )
         }
     }
+
+    @HandleEvent(onlyOnSkyblock = true)
+    fun onItemAdd(event: ItemAddEvent) {
+        if (event.amount != 1 || event.source != ItemAddManager.Source.ITEM_ADD) return
+        if (!config.enchantedBook) return
+        val internalName = event.internalName
+        val category = internalName.getItemStackOrNull()?.getItemCategoryOrNull() ?: return
+        if (category != ItemCategory.ENCHANTED_BOOK) return
+        if (inAnyIsland(ignoredBookIslands)) return
+
+        val anyRecent = ChatUtils.chatLines.none {
```

Comment:
when using patcher, this will iterate over many many lines. maybe first use `take()` to not process everything but only the last 50ish messages? (i assume the messages are in order in the list)